include/uk/plat/lcpu.h

```diff
@@ -1,104 +1,164 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UKPLAT_LCPU_H__
+#define __UKPLAT_LCPU_H__
+
+#include <uk/arch/time.h>
+#include <uk/essentials.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Enables interrupts
+ */
+void ukplat_lcpu_enable_irq(void);
+
+/**
+ * Disables interrupts
+ */
+void ukplat_lcpu_disable_irq(void);
+
+/**
+ * Returns current interrupt flags and disables them
+ * @return interrupt flags (Note that the format is unspecified)
+ */
+unsigned long ukplat_lcpu_save_irqf(void);
+
+/**
+ * Loads interrupt flags
+ * @param flags interrupt flags (Note that the format is unspecified)
+ */
+void ukplat_lcpu_restore_irqf(unsigned long flags);
+
+/**
+ * Checks if interrupts are disabled
+ * @return non-zero value if interrupts are disabled
+ */
+int ukplat_lcpu_irqs_disabled(void);
+
+void ukplat_lcpu_irqs_handle_pending(void);
+
+/**
+ * Halts the current logical CPU execution
+ */
+void ukplat_lcpu_halt(void);
+
+/**
+ * Halts the current logical CPU execution
+ * Execution is returned when an interrupt/signal arrived or
+ * the specified deadline expired
+ * @param until deadline in nanoseconds
+ */
+void ukplat_lcpu_halt_to(__snsec until);
+
+/**
+ * Halts the current logical CPU execution
+ * Execution is returned when an interrupt/signal arrived
+ */
+void ukplat_lcpu_halt_irq(void);
+
+#ifdef CONFIG_HAVE_SMP
+#include <uk/list.h>
+
+typedef void (*ukplat_lcpu_entry_t)(void) __noreturn;
+typedef __u32 __lcpuid;
+
+enum lcpu_states { offline, busy, idle, waiting };
+
+struct ukplat_lcpu_func {
+	struct uk_list_head lentry;
+	void (*fn)(struct __regs *regs, struct ukplat_lcpu_func *fn);
+	void *user;
+};
+
+struct ukplat_cpu {
+	ukplat_lcpu_entry_t entry;
+	void *stackp;
+	enum lcpu_states state;
+	struct uk_list_head fnlist;

```

Comment:
We could save a bit of memory by using a union. Something like this:
```C
struct x86_lcpu_startup_args {
	/* add field description */
	ukplat_lcpu_entry_t entry;
	/* add field description */			
	void *stackp;
};

struct x86_cpu {
	/* add comment why we can share memory */
	union {
		/* add field description */
		struct x86_lcpu_startup_args s_args;
		/* add field description */
		struct uk_list_head fn_list;
	};
	/* add field description */
	volatile enum x86_lcpu_states state;
};
```

Don't forget to initialize the list head using `UK_INIT_LIST_HEAD` after using the arguments.